crates/ra_ide_api/src/hover.rs

```diff
@@ -118,9 +177,50 @@ impl NavigationTarget {
         // TODO: After type inference is done, add type information to improve the output
         let node = self.node(db)?;
 
+        // FIXME: This is copied from `structure.rs` and should probably
+        // be moved somewhere common
+        fn collapse_ws(node: &SyntaxNode, output: &mut String) {
+            let mut can_insert_ws = false;
+            for line in node.text().chunks().flat_map(|chunk| chunk.lines()) {
+                let line = line.trim();
+                if line.is_empty() {
+                    if can_insert_ws {
+                        output.push_str(" ");
+                        can_insert_ws = false;
+                    }
+                } else {
+                    output.push_str(line);
+                    can_insert_ws = true;
+                }
+            }
+        }
+
+        // FIXME: This is also partially copied from `structure.rs`
+        fn visit_fn(node: &ast::FnDef) -> Option<String> {
```

Comment:
`collapse_ws` is directly copied from `structure.rs` and `visit_fn` is very similiar to the functionality in `structure.rs` so I'd like to refactor and move at least `collapse_ws` to some common place.